### PR TITLE
[MM-51737] Cluster Installation Status endpoint

### PR DESCRIPTION
**Pull request by mirshahriar**
#### Summary

This endpoint `/api/cluster_installation/%s/status` will return the status of running the mm servers.

```
type ClusterInstallationStatus struct {
	InstallationFound bool   `json:"installation_found"`
	Replicas          *int32 `json:"replicas,omitempty"`
	TotalPod          *int32 `json:"total_pod,omitempty"`
	RunningPod        *int32 `json:"running_pod,omitempty"`
	ReadyPod          *int32 `json:"ready_pod,omitempty"`
	StartedPod        *int32 `json:"started_pod,omitempty"`
	ReadyLocalServer  *int32 `json:"ready_local_server,omitempty"`
}
```

The `cloud` command `cloud cluster installation status --cluster-installation=` will fetch from that endpoint.

#### Ticket Link

  Fixes https://mattermost.atlassian.net/browse/MM-51737

#### Release Note

```release-note
None
```

### CODE REVIEW

File: model/cluster_installation.go [hunk 1]
```diff
@@ -33,6 +35,16 @@ type ClusterInstallationFilter struct {
...
 
+type ClusterInstallationStatus struct {
+	InstallationFound bool   `json:"installation_found"`
+	Replicas          *int32 `json:"replicas,omitempty"`
+	TotalPod          *int32 `json:"total_pod,omitempty"`
+	RunningPod        *int32 `json:"running_pod,omitempty"`
+	ReadyPod          *int32 `json:"ready_pod,omitempty"`
+	StartedPod        *int32 `json:"started_pod,omitempty"`
+	ReadyLocalServer  *int32 `json:"ready_local_server,omitempty"`
+}
```
Comment (gabrieljackson) [inline model/cluster_installation.go:46]: Historically, we have used the non-underline JSON naming scheme for our various structs: https://github.com/mattermost/mattermost-cloud/blob/master/model/group_status.go#L12-L20

This wasn't really an intentional decision, but sort of became the standard for the codebase. Do you think we should change that going forward or keep the original style. @fmartingr have any thoughts as well?

Comment (mirshahriar) [reply model/cluster_installation.go:46]: Let's stick with the initial design. If we want to change, we also need to consider how to handle DB objects.

File: e2e/workflow/installation.go [hunk 2]
```diff
@@ -117,6 +117,10 @@ func (w *InstallationSuite) GetCI(ctx context.Context) error {
 	return nil
 }
 
+func (w *InstallationSuite) CheckClusterInstallationStatus(ctx context.Context) error {
+	return pkg.WaitForClusterInstallationReadyStatus(w.client, w.Meta.ClusterInstallationID, w.logger)
```
Comment (fmartingr) [inline e2e/workflow/installation.go:121]: Maybe we can merge `GetCI` and `CheckClusterInstallationStatus` into the  same step?

File: internal/provisioner/cluster_installation_provisioner.go [hunk 3]
```diff
@@ -522,6 +524,86 @@ func (provisioner Provisioner) getMattermostCustomResource(cluster *model.Cluste
...
+
+	var status model.ClusterInstallationStatus
+
+	deployment, err := k8sClient.Clientset.AppsV1().Deployments(clusterInstallation.Namespace).Get(context.TODO(), makeClusterInstallationName(clusterInstallation), metav1.GetOptions{})
+	if err != nil {
+		if !k8sErrors.IsNotFound(err) {
+			return nil, errors.Wrap(err, "failed to query mattermost deployment")
+		}
+
+		status.InstallationFound = false
```
Comment (fmartingr) [inline internal/provisioner/cluster_installation_provisioner.go:541]: This should default to `false`.

File: model/cluster_installation.go (hunk 1 above)
Comment (fmartingr) [reply model/cluster_installation.go:46]: My personal preference is `snake_case` for APIs, but I'd rather stick with `CamelCase` in the provisioner than switch everything to snakes, I'd rather be homogeneus than picky about this. 1/5

File: model/cluster_installation.go [hunk 4]
```diff
@@ -124,3 +136,13 @@ func MigrateClusterInstallationResponseFromReader(reader io.Reader) (*MigrateClu
 
 	return &migrateClusterInstallationResponse, nil
 }
+
+func NewClusterInstallationStatusFromReader(reader io.Reader) (*ClusterInstallationStatus, error) {
```
Comment (fmartingr) [inline model/cluster_installation.go:140]: Shouldn't we get a lint error when a public function does not have a comment? cc: @gabrieljackson

**Review by fmartingr — APPROVED** [review]
Looks good to me. Ingenius way to check the server status by the way.

While I agree this make our e2e more robust, I'd say we may have found a _something_ in either the operator or the server. We should trust that our ready pods are truly ready, so if the server is reporting that is ready when it could not be the case this could lead to other problems in the future. So the end goal here should be not requiring to manually check if the server is properly running since it should report that itself correctly.

File: e2e/workflow/installation.go (hunk 2 above)
Comment (mirshahriar) [reply e2e/workflow/installation.go:121]: In that case, do we need a better name than `GetCI`?

File: model/cluster_installation.go (hunk 4 above)
Comment (mirshahriar) [reply model/cluster_installation.go:140]: We need to configure that in linters. Looks like it's missing.

Comment (gabrieljackson) [reply model/cluster_installation.go:140]: Huh, that used to be working. We must have lost it along the way somehow.

**Review by gabrieljackson — APPROVED** [review]
👍

File: e2e/workflow/installation.go (hunk 2 above)
Comment (mirshahriar) [reply e2e/workflow/installation.go:122]: I'm leaving it now as it it.

**Pull request merged**